mbrl/algorithms/dreamer.py

```diff
@@ -0,0 +1,221 @@
+    # to simulate the trajectories from the prior transition model
+    # The starting point for trajectories is each imagined state output by the
+    # representation model from the dataset of environment observations
+    agent: DreamerAgent = create_dreamer_agent_for_model(
+        planet, model_env, cfg.algorithm.agent
+    )
+
+    # Callback and containers to accumulate training statistics and average over batch
+    rec_losses: List[float] = []
+    reward_losses: List[float] = []
+    policy_losses: List[float] = []
+    critic_losses: List[float] = []
+    kl_losses: List[float] = []
+    model_grad_norms: List[float] = []
+    agent_grad_norms: List[float] = []
+
+    def get_metrics_and_clear_metric_containers():
+        metrics_ = {
+            "observations_loss": np.mean(rec_losses).item(),
+            "reward_loss": np.mean(reward_losses).item(),
+            "policy_loss": np.mean(policy_losses).item(),
+            "critic_loss": np.mean(critic_losses).item(),
+            "model_gradient_norm": np.mean(model_grad_norms).item(),
+            "agent_gradient_norm": np.mean(agent_grad_norms).item(),
+            "kl_loss": np.mean(kl_losses).item(),
+        }
+
+        for c in [
+            rec_losses,
+            reward_losses,
+            policy_losses,
+            critic_losses,
+            kl_losses,
+            model_grad_norms,
+            agent_grad_norms,
+        ]:
+            c.clear()
+
+        return metrics_
+
+    def model_batch_callback(_epoch, _loss, meta, _mode):
+        if meta:
+            rec_losses.append(meta["observations_loss"])
+            reward_losses.append(meta["reward_loss"])
+            kl_losses.append(meta["kl_loss"])
+            if "grad_norm" in meta:
+                model_grad_norms.append(meta["grad_norm"])
+
+    def agent_batch_callback(_epoch, _loss, meta, _mode):
+        if meta:
+            policy_losses.append(meta["policy_loss"])
+            critic_losses.append(meta["critic_loss"])
+            if "grad_norm" in meta:
+                agent_grad_norms.append(meta["grad_norm"])
+
+    def is_test_episode(episode_):
+        return episode_ % cfg.algorithm.test_frequency == 0
+
+    # Dreamer loop
+    step = replay_buffer.num_stored
+    total_rewards = 0.0
+    for episode in tqdm(range(cfg.algorithm.num_episodes)):
+        # Train the model for one epoch of `num_grad_updates`
+        dataset, _ = get_sequence_buffer_iterator(
+            replay_buffer,
+            cfg.overrides.batch_size,
+            0,  # no validation data
+            cfg.overrides.sequence_length,
+            max_batches_per_loop_train=cfg.overrides.num_grad_updates,
+            use_simple_sampler=True,
+        )
+        trainer.train(
+            dataset, num_epochs=1, batch_callback=model_batch_callback, evaluate=False
+        )
+        agent.train(dataset, num_epochs=1, batch_callback=agent_batch_callback)
```

Comment:
> I'm wondering if we should be passing a different iterator for training the agent. If I understand the paper correctly, the Dreamer agent is trained on trajectories whose start states are sampled from the experience buffer, but where all subsequent states are obtained by rolling out the model. In this case, we only need to sample individual transitions to get start states, and not full sequences, which is what `dataset` would return here.

I might have misunderstood the paper, but I'm not sure this is correct. In Algorithm 1 (Page 3), they:
1. Draw  `B` data sequences or episodes `{(a_t, o_t, r_t)} ~_{t=k}^{k+L}`. Here `k` is the outer variable looping over episodes, while `t` is the inner variable looping over timesteps in an episode.
2. Compute model states `s_t` for all `t` in `[k, k + L)` for all `k` in `B` using the RSSM transition model.
3. Imagine trajectories `{s_\tau, a_\tau}_{\tau = t}^{\tau = t + H}` from each state `s_t` in B, *not* just the initial state `s_k` in each episode.

I'm not sure if this explanation was clear, and I'll take another look at the prior implementations linked in the other comment to confirm.

Reply:
Looking at the paper again, I think your interpretation is correct because the `Compute model states` step occurs for all sampled `o_t`, and then trajectories are imagined for all model states `s_t`. I find it a bit confusing how they are using the index `k`; I guess this increasing in increments of size `L`? That is, the j-th trajectory goes from `t=L*(j-1)` to `L*j - 1`? In any case, confirming with prior implementations is a good idea. 

Regarding the performance hit, one idea that wouldn't require a lot of changes would be to add get/set methods for random state of the iterator, so that we can have it return the same set of samples both for the model and agent loops. We should then be able to use the model trainer callback to store all computed model states, and pass them to the agent trainer in the correct order. 

Does that make sense?